python/lsst/cp/pipe/utils.py

```diff
@@ -710,6 +711,415 @@ def arrangeFlatsByExpId(exposureList, exposureIdList):
+        w2 = weightImage[-dy:, :nRows - dx]
+    # use the same mask for all 3 calculations
+    wAll = w1*w2
+    # do not use mean() because weightImage=0 pixels would then count
+    nPix = wAll.sum()
+    im1TimesW = im1*wAll
+    s1 = im1TimesW.sum()/nPix
+    s2 = (im2*wAll).sum()/nPix
+    p = (im1TimesW*im2).sum()/nPix
+    cov = p - s1*s2
+
+    return cov, nPix
+
+
+def getFitDataFromCovariances(i, j, mu, fullCov, fullCovModel, fullCovSqrtWeights, gain=1.0,
+                              divideByMu=False, returnMasked=False):
+    """Get measured signal and covariance, cov model, weigths, and mask at
+    covariance lag (i, j).
+
+    Parameters
+    ----------
+    i :  `int`
+        Lag for covariance matrix.
+
+    j : `int`
+        Lag for covariance matrix.
+
+    mu : `list`
+        Mean signal values.
+
+    fullCov : `list` of `numpy.array`
+        Measured covariance matrices at each mean signal level in mu.
+
+    fullCovSqrtWeights : `list` of `numpy.array`
+        List of square root of measured covariances at each mean
+        signal level in mu.
+
+    fullCovModel : `list` of `numpy.array`
+        List of modeled covariances at each mean signal level in mu.
+
+    gain : `float`, optional
+        Gain, in e-/ADU. If other than 1.0 (default), the returned
+        quantities will be in electrons or powers of electrons.
+
+    divideByMu : `bool`, optional
+        Divide returned covariance, model, and weights by the mean
+        signal mu?
+
+    returnMasked : `bool`, optional
+        Use mask (based on weights) in returned arrays (mu,
+        covariance, and model)?
+
+    Returns
+    -------
+    mu : `numpy.array`
+        list of signal values at (i, j).
+
+    covariance : `numpy.array`
+        Covariance at (i, j) at each mean signal mu value (fullCov[:, i, j]).
+
+    covarianceModel : `numpy.array`
+        Covariance model at (i, j).
+
+    weights : `numpy.array`
+        Weights at (i, j).
+
+    maskFromWeights : `numpy.array`, optional
+        Boolean mask of the covariance at (i,j), where the weights
+        differ from 0.
+
+    Notes
+    -----
+    This function is a method of the `CovFit` class.
+    """
+    mu = np.array(mu)
+    fullCov = np.array(fullCov)
+    fullCovModel = np.array(fullCovModel)
+    fullCovSqrtWeights = np.array(fullCovSqrtWeights)
+    covariance = fullCov[:, i, j]*(gain**2)
+    covarianceModel = fullCovModel[:, i, j]*(gain**2)
+    weights = fullCovSqrtWeights[:, i, j]/(gain**2)
+
+    maskFromWeights = weights != 0
+    if returnMasked:
+        weights = weights[maskFromWeights]
+        covarianceModel = covarianceModel[maskFromWeights]
+        mu = mu[maskFromWeights]
+        covariance = covariance[maskFromWeights]
+
+    if divideByMu:
+        covariance /= mu
+        covarianceModel /= mu
+        weights *= mu
+    return mu, covariance, covarianceModel, weights, maskFromWeights
+
+
+def makeCovArray(inputTuple, maxRangeFromTuple=8):
```

Comment:
Is this ever called outside of PhotonTransferCurveExtractTask?  I have the same question with `computeCovDirect` above.

Reply:
I think it's only called by the extract class, so I moved it there.